subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -59,33 +59,185 @@ static struct lwm2m_engine_obj_inst inst;
 #endif
 
+u8_t lwm2m_firmware_get_update_state(void)
+{
+	return update_state;
+}
+
+void lwm2m_firmware_set_update_state(u8_t state)
+{
+	bool error = false;
+
+	/* Check LWM2M SPEC appendix E.6.1 */
+	switch (state) {
+	case STATE_DOWNLOADING:
+		if (update_state != STATE_IDLE) {
+			error = true;
+		}
+		break;
+	case STATE_DOWNLOADED:
+		if (update_state != STATE_DOWNLOADING &&
+		    update_state != STATE_UPDATING) {
+			error = true;
+		}
+		break;
+	case STATE_UPDATING:
+		if (update_state != STATE_DOWNLOADED) {
+			error = true;
+		}
+		break;
+	case STATE_IDLE:
+		break;
+	default:
+		SYS_LOG_ERR("Unhandled state: %u", state);
+		return;
+	}
+
+	if (error) {
+		SYS_LOG_ERR("Invalid state transition: %u -> %u",
+			    update_state, state);
+	}
+
+	update_state = state;
+	NOTIFY_OBSERVER(LWM2M_OBJECT_FIRMWARE_ID, 0, FIRMWARE_STATE_ID);
+	SYS_LOG_DBG("Update state = %d", update_state);
+}
+
+u8_t lwm2m_firmware_get_update_result(void)
+{
+	return update_result;
+}
+
+void lwm2m_firmware_set_update_result(u8_t result)
+{
+	u8_t state;
+	bool error = false;
+
+	/* Check LWM2M SPEC appendix E.6.1 */
+	switch (result) {
+	case RESULT_DEFAULT:
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_SUCCESS:
+		if (update_state != STATE_UPDATING) {
+			error = true;
+			state = update_state;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_NO_STORAGE:
+	case RESULT_OUT_OF_MEM:
+	case RESULT_CONNECTION_LOST:
+	case RESULT_UNSUP_FW:
+	case RESULT_INVALID_URI:
+	case RESULT_UNSUP_PROTO:
+		if (update_state != STATE_DOWNLOADING) {
+			error = true;
+			state = update_state;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_INTEGRITY_FAILED:
+		if (update_state != STATE_DOWNLOADING &&
+		    update_state != STATE_UPDATING) {
+			error = true;
+			state = update_state;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_UPDATE_FAILED:
+		if (update_state != STATE_UPDATING) {
+			error = true;
+			state = update_state;
+		}
+
+		/* Next state could be idle or downloaded */
+		break;
+	default:
+		SYS_LOG_ERR("Unhandled result: %u", result);
+		return;
+	}
+
+	if (error) {
+		SYS_LOG_ERR("Unexpected result(%u) set while state is %u",
+			    result, state);
+	}
+
+	update_result = result;
+	NOTIFY_OBSERVER(LWM2M_OBJECT_FIRMWARE_ID, 0, FIRMWARE_UPDATE_RESULT_ID);
+	SYS_LOG_DBG("Update result = %d", update_result);
+}
+
 static int package_write_cb(u16_t obj_inst_id,
 			    u8_t *data, u16_t data_len,
 			    bool last_block, size_t total_size)
 {
-	SYS_LOG_DBG("PACKAGE WRITE");
+	u8_t state;
+	int ret = 0;
+
+	state = lwm2m_firmware_get_update_state();
+	if (state == STATE_IDLE) {
+		/* TODO: setup timer to check download status,
+		 * make sure it fail after timeout
+		 */
+		lwm2m_firmware_set_update_state(STATE_DOWNLOADING);
+	} else if (state != STATE_DOWNLOADING) {
+		if (data_len == 0 && state == STATE_DOWNLOADED) {
```

Comment:
I wonder if we should always check for data_len == 0 even when state is not downloading. That is just to protect the code flow as the state can be updated via other parts of the code (even when it shouldn't).

Reply:
Just to confirm. Are you suggesting that we always check if "data_len == 0" and reset the state back to default?

Reply:
Yes, my concern is what would happen when data_len == 0 but state is not what we expect to  be (simply because state can be changed independently).